Implement tokio support

jmagnuson opened the pull request:
Adds dependencies `tokio` + `futures` behind 'tokio' feature.

Implements `tokio::UnixSeqpacketConn` and `tokio::UnixSeqpacketListener` for supporting `SEQPACKET` sockets in Tokio.

Closes #1.

---
Much of this roughly follows how tokio [`UnixStream`](https://docs.rs/tokio/0.2.22/tokio/net/struct.UnixStream.html)/`Listener`s (`SOCK_STREAM`) are set up, but attempts to adopt some of the existing API naming used by `uds`. However, I'm totally open to other naming conventions, or module organization for that matter.

My immediate use-case is for `UnixSeqPacketConn`, but have also introduced (perhaps a bit more roughly) the Listener type for the sake of some sort of completeness.

jmagnuson commented on Cargo.toml:
```diff
@@ -25,7 +29,12 @@ mio_07 = {package="mio", version="0.7.0", features=["os-util", "uds"], optional=
+
+[target."cfg(unix)".dev-dependencies]
+tempdir = "0.3"
+tokio_02 = {package="tokio", version = "0.2", features = ["macros", "rt-core", "test-util"]}
```
Ideally this would be `optional` as well, but I don't think Cargo allows optional dev-dependencies. Thus, it's either forcing these tokio test-related features for non-test builds, or forcing all test builds to take a tokio dependency.

tormol replied:
What do you think about moving the dev dependency features to RUSTFLAGS? If it's not pointless I can do that later.

jmagnuson replied:
Interesting; I wasn't aware of this issue, or workaround. I tried moving the dev features to `RUSTFLAGS`, but had some issues with `macros` and tokio's failure to pull in the `tokio-macros` sub-crate:

```
$ RUSTFLAGS='--cfg feature="os-poll" --cfg feature="rt-core" --cfg feature="macros"' cargo test --all-features

...

error[E0432]: unresolved import `tokio_macros`
   --> /home/jon/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.22/src/lib.rs:403:13
    |
403 |     pub use tokio_macros::select_priv_declare_output_enum;
    |             ^^^^^^^^^^^^ use of undeclared type or module `tokio_macros`
```

(note, I realized that `test-util` isn't actually used, so that will be dropped regardless of the outcome of this)

One workaround is just importing `tokio-macros` itself into dev-dependencies, and then setting up tests such that:

```rust
#[tokio_macros::test]
async fn test_listener_accept() {
    ...
}
```

which isn't something I've really seen other crates do (especially in documentation, which exclusively uses `tokio::main`), but I couldn't come up with a better solution offhand.

Thoughts?

jmagnuson replied:
There's also [`cargo-hack`](https://github.com/taiki-e/cargo-hack) which looks to address these limitations in cargo. I did a quick trial run and it didn't seem to like running test with `--all-features`, but worked as expected when I wrote everything out manually:

```
$ cargo hack --features="mio_07,tokio,tokio_02/macros,tokio_02/rt-core,mio_07/os-poll" test
```

tormol replied:
Thanks for the attempt. We'll just have to accept it being enabled for now.
(by the time I'm releasing v1.0 of this it will probably have been fixed in cargo :)

cargo-hack looks interesting but feels too complex.

jmagnuson commented:
I'll add notes to the README and determine what needs to be done to satisfy OSX CI, but any interim reviews would be appreciated.

jmagnuson commented on src/tokio/seqpacket.rs:
```diff
@@ -0,0 +1,200 @@
+        Ok(conn)
+    }
+
+    pub(crate) fn new(conn: nonblocking::UnixSeqpacketConn) -> io::Result<UnixSeqpacketConn> {
```
It would actually be nice to expose this or `impl TryFrom` to gain some composition with the other [non]-blocking types. For instance, I'd really like to take `uds::UnixSeqpacketConn::pair()` and somehow convert them into the tokio variants.

jmagnuson replied:
I had the same thought right after commenting-- mainly just to reduce API friction by having different `pair` functions and avoiding indirection. But yeah, there's no point in forcing additional syscalls on the user. I'll add `from_nonblocking` as well as `pair`.

jmagnuson replied:
Done in 6ed49031f9cd5473e9b8be582716d06ae58e1c97.

tormol reviewed (changes requested):
Thanks!

You probably also need to update the MSRV to what tokio or futures require.
(The next uds release will be a major version due to other changes anyway).

Do you plan to add more methods? If not I'll add `bind_unix_addr()` and the like later.

jmagnuson commented on src/tokio/seqpacket.rs:
```diff
@@ -0,0 +1,200 @@
+    }
+}
+
+impl AsyncWrite for UnixSeqpacketConn {
```
I touch upon the reasoning for this in the [notes](https://github.com/tormol/uds/pull/2/files#diff-145eba58ca323c6c0ec060ca120c3ebbR3-R9), but it is definitely a point of discussion, and some contention especially surrounding the codec/`Framed` API and resulting functionality like `Sink` and `Stream`.

It seems that tokio's `UnixDatagram` implementation (which serves as the basis for this comparison because of its similarity to `Seqpacket`) has over time gotten less attention than its `UnixStream` counterpart. My coworker actually [introduced `UnixDatagramFramed`](https://github.com/tokio-rs/tokio/pull/453) since, similar to your point, `AsyncRead`/`Write` doesn't really make sense in the context of datagrams, so a completely separate API was required for codec support. But as far as I can tell, the tokio 0.2 migration did not carry over this functionality, perhaps due to lack of interest, and I suspect that any implementation could be subject to further breakage as tokio 0.3, 1.0, etc. are forthcoming.

This leads the discussion to `Seqpacket`-- I took a bit of a hand-wavy shortcut by saying that since sequenced packets behave mostly like unix streams, then the same API could be applied with the understanding, at the application layer, that sender and receiver should have some mutual understanding of the packet boundaries. This allows for tapping into useful, composable features like codec, `Stream[Ext]`, etc. for "free" since everything is already built for stream-type sockets. That said, I'm happy to go the `Datagram` route of having `UnixSeqpacketConnFramed`, and then impl'ing `Sink` and `Stream` on top of that. There does at least exist a [`UdpFramed`](https://docs.rs/tokio-util/0.3.1/tokio_util/udp/struct.UdpFramed.html) type whose own decoupled API could be modeled after.

jmagnuson replied:
I brought up this potential solution in the tokio discord, and the feedback wasn't very positive. And I also learned that `UdpFramed` going away, so I think I'll just drop these trait impl's, and punt on any kind of codec support in this PR.

jmagnuson commented on src/tokio/seqpacket.rs:
```diff
@@ -0,0 +1,200 @@
+    }
+
+    fn poll_shutdown(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        // self.shutdown(std::net::Shutdown::Write)?;
```
Ah, yes. Socket shutdown gets propagated to existing tokio/mio-uds sockets via, e.g. [`UnixStream::shutdown`](https://doc.rust-lang.org/nightly/std/os/unix/net/struct.UnixStream.html#method.shutdown), but a similar function doesn't yet exist for `UnixSeqpacketConn`. So that will need to get implemented first and then used to build this.

jmagnuson replied:
Blocking/mio `shutdown()` done in 4d6c2c9, and tokio wrapper 6ed4903.

tormol replied:
Thanks

jmagnuson commented on src/tokio/seqpacket.rs:
```diff
@@ -0,0 +1,200 @@
+        let conn = nonblocking::UnixSeqpacketConn::connect(path)?;
+        let conn = UnixSeqpacketConn::new(conn)?;
+
+        //poll_fn(|cx| conn.io.poll_write_ready(cx)).await?;
```
I think we want this actually-- having this hook will ensure that the socket is ready to be used once returned. Tokio's `UnixStream` and `TcpStream` follow a similar line of [reasoning](https://docs.rs/tokio/0.2.22/src/tokio/net/tcp/stream.rs.html#146-147), and my understanding is that `SOCK_SEQPACKET` connections behave similarly to `SOCK_STREAM`.

jmagnuson replied:
Done in 6ed49031f9cd5473e9b8be582716d06ae58e1c97.

jmagnuson commented:
I went ahead and rebased the `shutdown` change (4d6c2c9) to go before everything else, resulting in a force-push. However, I left the existing changeset otherwise in-tact so that the subsequent fixup commits could be reviewed.

>Do you plan to add more methods?

I think it might be best to leave that out for this initial pass. Not trying to push work onto you, but honestly you would probably get the tokio sockets up to parity a lot faster than I could. There's still the larger question of what to do about `AsyncRead`/`Write` traits that this PR should nail down first. And... making CI happy :(

jmagnuson commented on src/tokio/seqpacket.rs:
```diff
@@ -0,0 +1,200 @@
+    }
+}
+
+#[cfg(test)]
```
Also done in 6ed49031f9cd5473e9b8be582716d06ae58e1c97.

tormol commented on README.md:
```diff
@@ -88,10 +88,21 @@ Mio 0.7 is also supported:
 
-The minimum Rust version is 1.36, because of `std::io::IoSlice`.
-If this is a problem I can make the parts that need it opt-out.
+The minimum Rust version is 1.39, because of the MSRV of both `tokio` and `futures`.
```
Since those dependencies are optional, this makes it unclear what the minimum supported version is when not enabling the tokio feature. Either don't mention the reason or add that the minimum version is 1.36 by default.

tormol commented on src/seqpacket.rs:
```diff
@@ -650,6 +662,17 @@ impl NonblockingUnixSeqpacketConn {
     }
+
+    /// Shuts down the read, write, or both halves of this connection.
+    pub fn shutdown(&self, how: Shutdown) -> io::Result<()> {
```
Thanks for adding `shutdown()`!

Maybe add a test? (I can do that later though)

tormol reviewed (commented):
Thanks for rebasing.
After changing the MSRV documentation, you can squash the later commits into one commit adding stream and another adding listener.

Sorry for being slow to respond. I was traveling last weekend and have been working late.

tormol commented:
Forgot to say that the failing test is probably flaky or maybe apple changed something.

jmagnuson commented:
Went ahead and rebased again to clean things up since there are only one or two remaining items. But the latest changes were:

- added `shutdown` tests for blocking, nonblocking, and tokio socket variants.
- removed `AsyncRead`/`Write` impls
- reworded MSRV to exclude mentioning tokio
- removed `test-util` feature inclusion from tokio

Let me know your thoughts how what to do about `RUSTFLAGS` and tokio's `macros` feature, or if you want to punt and figure out what works best at a later time.

Sidenote: you may be right about CI flakiness. I see the latest commit c48d0af has passed, but I didn't make any explicit changes for that to happen..

jmagnuson commented:
@tormol any further thoughts? Tokio is also on the verge of releasing 0.3 (which will invariably resemble 1.0 more closely than 0.2) so that is something to consider.

tormol commented:
THANK YOU.

Sorry for being so slow to respond again.
(I lacked motivation to think about code outside of work for a while, but kept thinking I'd do it in a few days).
Now that I'm likely finished looking for an apartment to buy that will hopefully change.)

jmagnuson commented:
No worries at all-- know that feeling all too well, and I actually just had my first child in September, so my OSS work has been pretty inconsistent as well. Glad to see the PR make it across the finish line though!

tormol commented:
Congratulations!

I'm hoping to test this on NetBSD and IllumOS and release this weekend.
Before doing that I'm going to try to stabilize the macOS CI and commit some features that libc now supports.


Den 16. oktober 2020 00:49:52 CEST, skrev Jon Magnuson <notifications@github.com>:
>No worries at all-- know that feeling all too well, and I actually just
>had my first child in September, so my OSS work has been pretty
>inconsistent as well. Glad to see the PR make it across the finish line
>though!
>
>-- 
>You are receiving this because you modified the open/close state.
>Reply to this email directly or view it on GitHub:
>https://github.com/tormol/uds/pull/2#issuecomment-709630283

tormol commented:
0.2.0 is now released.

When I tested on Illumos, the test binaries created by `cargo test --features tokio` failed to start due to a missing symbol. I've just noted that in a feature table I've added to the README, but it could just be because I haven't updated that VM (OmniOS) in a while.
